File: java/testsuite/Jogl4Examples/MeshExample/src/gui/MeshKeyboardInteractionTechniques.java
package gui;

import model.MeshModel;
import vsdk.toolkit.gui.KeyEvent;
import vsdk.toolkit.gui.CameraController;
import vsdk.toolkit.gui.RendererConfigurationController;

public class MeshKeyboardInteractionTechniques {
    private final MeshModel model;
    private final CameraController cameraController;
    private final RendererConfigurationController qualityController;

    public MeshKeyboardInteractionTechniques(
        MeshModel model,
        CameraController cameraController,
        RendererConfigurationController qualityController)
    {
        this.model = model;
        this.cameraController = cameraController;
        this.qualityController = qualityController;
    }

    public boolean processKeyPressedEvent(KeyEvent event)
    {
        if ( event == null ) {
            return false;
        }

        if ( event.keycode == KeyEvent.KEY_ESC ) {
            System.exit(0);
        }
        if ( event.keycode == KeyEvent.KEY_I ) {
            System.out.println(model.getQualitySelection());
            return true;
        }
        if ( cameraController.processKeyPressedEvent(event) ) {
            return true;
        }
        if ( qualityController.processKeyPressedEvent(event) ) {
            System.out.println(model.getQualitySelection());
            return true;
        }

        return false;
    }

    public boolean processKeyReleasedEvent(KeyEvent event)
    {
        if ( event == null ) {
            return false;
        }

        if (cameraController.processKeyReleasedEvent(event)) {
            return true;
        }
        if (qualityController.processKeyReleasedEvent(event)) {
            return true;
        }
        return false;
    }
}

File: cpp/testsuite/OpenGL4Examples/MeshExample/src/gui/MeshKeyboardInteractionTechniques.cpp
#include <cstdio>

#include "vsdk/toolkit/gui/CameraController.h"
#include "vsdk/toolkit/gui/KeyEvent.h"
#include "vsdk/toolkit/gui/RendererConfigurationController.h"
#include "gui/MeshKeyboardInteractionTechniques.h"
#include "model/MeshModel.h"

MeshKeyboardInteractionTechniques::MeshKeyboardInteractionTechniques(
    MeshModel* model,
    CameraController* cameraController,
    RendererConfigurationController* qualityController,
    bool* shouldClose)
    : model(model),
      cameraController(cameraController),
      qualityController(qualityController),
      shouldClose(shouldClose)
{
}

bool MeshKeyboardInteractionTechniques::processKeyPressedEvent(const KeyEvent& event)
{
    if ( event.keycode == KeyEvent::KEY_ESC ) {
        if ( shouldClose != 0 ) {
            *shouldClose = true;
        }
        return true;
    }
    if ( event.keycode == KeyEvent::KEY_I ) {
        std::printf("%s\n", model->getQualitySelection()->toString().c_str());
        return true;
    }
    if ( cameraController != 0 && cameraController->processKeyPressedEvent(event) ) {
        return true;
    }
    if ( qualityController != 0 && qualityController->processKeyPressedEvent(event) ) {
        std::printf("%s\n", model->getQualitySelection()->toString().c_str());
        return true;
    }

    return false;
}

bool MeshKeyboardInteractionTechniques::processKeyReleasedEvent(const KeyEvent& event)
{
    if ( cameraController != 0 && cameraController->processKeyReleasedEvent(event) ) {
        return true;
    }
    if ( qualityController != 0 && qualityController->processKeyReleasedEvent(event) ) {
        return true;
    }
    return false;
}
